src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java

```diff
@@ -363,16 +363,33 @@ private void updateUninitializedVariables(Deque<DetailAST> prevScopeUnitializedV
     }
 
     /**
-     * If token is LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, or LITERAL_ELSE, then do not
-     * update the uninitialized variables.
+     * If token is LITERAL_IF and there is an {@code else} following or token is CASE_GROUP and
+     * there is another {@code case} following, then update the uninitialized variables.
      * @param ast token to be checked
      * @return true if should be updated, else false
      */
     private static boolean shouldUpdateUninitializedVariables(DetailAST ast) {
-        return ast.getType() != TokenTypes.LITERAL_TRY
-                && ast.getType() != TokenTypes.LITERAL_CATCH
-                && ast.getType() != TokenTypes.LITERAL_FINALLY
```

Comment:
In fact they are not removed, on the contrary the condition exp is extended.
The original code used `!=` and `&&` and the new code is using `==` and `||`.
To make it clearer, in fact, the possible type of the `ast` could be one of the following: 
- try
- else
- catch
- finally
- if
- method def
- lambda
- synchronized
- for
- while
- do
- case group
- { (SLIST)
- constructor def
- static init
- instance init

The original code means that if the ast type is not one of the first four (try, catch, finally, else), then it should update. And current code means that if the ast type is not one of try, catch, finally, method def, lambda ... then it should be update.

To find out the purpose of this method, before fixing this issue, I reviewed PR https://github.com/checkstyle/checkstyle/pull/2814, before which is merged there were no violation and after merged the false positive appears. It is sure there is smth wrong in that commit.
That PR is to fix https://github.com/checkstyle/checkstyle/issues/2807, to assert violation on smth like this:
```
int a;
if (true) {
  a = 1; 
} // (1)
else { // (2) 
  a = 2; // (3)
} // (4)
```
In that PR, method `shouldUpdateUninitializedVariables` was added to confirm that we will update `a` into the uninitialized variables list when we leave token at (1), and will not remove `a` from violation candidates list when we are processing at (3)
Therefore we should only do the update when we leave a if scope which has an else scope following or a case group witch has another case group following.
If we are leaving the scope of smth like try, catch ... we shouldn't do the update at all obviously. But there are many other type like try, catch, finally that is not taken into judgement and that why the false positive occurs.
Actually smth like this can reproduce false positive:
```
int a;
if (true) {}
else {
  a = 1;
  {} // (5)
  a = 2; (6)
}
```
It is because we do the update when leaving (5) and do not remove `a` from candidates list when processing (6). And actually we shouldn't do the update.

\-\-\-
Regrading the test. I don't understand clearly. I add the code mentioned in the issue into the input file. What kind of UT should I add for my changes? I don't have an idea. Could you please give me some more tips?